### PR TITLE
Fixed jobstop() to return an exit status

### PR DESCRIPTION
**Pull request by dexterfhy**
https://github.com/neovim/neovim/issues/10943

Hey, this is my first contribution to a project and was hoping if I could get feedback on whether my changes were in the right direction. Thank you!

### CODE REVIEW

File: src/nvim/eval.c [hunk 1]
```diff
@@ -12393,36 +12393,40 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 }
 
 // "jobstop()" function
-static void f_jobstop(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+// Returns 0, or -1 on error
```
Comment (justinmk) [inline src/nvim/eval.c]: all of the `f_xx` functions are documented in `:help eval.txt` rather than in docstrings (although it would be nice if eval.txt were generated from these docstrings, but until someone implements that, keep it in one place ...)

File: src/nvim/eval.c [hunk 2]
```diff
@@ -12393,36 +12393,40 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
...
+      status = -1;
+    }
   }
   process_stop((Process *)&data->stream.proc);
   rettv->vval.v_number = 1;
-  if (error) {
-    EMSG(error);
-  }
+
+  return status;
```
Comment (justinmk) [inline src/nvim/eval.c]: `f_xx` functions are VimL functions (`:help eval.txt`). They return things in `rettv`. See other functions for examples. Try out your changes by `:echo jobstop()` after building. (Or better, write a test in `job_spec.lua`.

File: src/nvim/eval.c [hunk 3]
```diff
@@ -12393,36 +12393,40 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
...
+static int f_jobstop(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 {
   rettv->v_type = VAR_NUMBER;
   rettv->vval.v_number = 0;
 
+  int status = 0;  // zero or negative error code
+
   if (check_restricted() || check_secure()) {
-    return;
+    return status;
```
Comment (justinmk) [inline src/nvim/eval.c]: this isn't needed/wanted because `rettv->vval.v_number = 0` already.

File: src/nvim/eval.c (hunk 2 above)
Comment (dexterfhy) [reply src/nvim/eval.c]: Hi Justin ) I appreciate you taking the time to give me detailed feedback. Meanwhile I'm still trying to build it on Windows but I'm encountering a couple of errors.

I referred to `f_jobstart()`. Does this mean that the return value is set via `rettv->vval.vnumber`? 

I've removed the comments I inserted, and modified rettv->vval.v_number to -1 if an error message is assigned during `channel_close` instead. Is this the right approach?

Comment (justinmk) [reply src/nvim/eval.c]: is there a reason to use -1 instead of 0?

> Does this mean that the return value is set via `rettv->vval.vnumber`?

yes. Getting it to build will be helpful so you can try the changes (and write a test). Perhaps create a ticket with the build log, maybe someone can figure it out.

Comment (dexterfhy) [reply src/nvim/eval.c]: My initial thought was to differentiate the status from the other errors, but I suppose not. Anyway, I managed to build and run it on an Ubuntu VM, so I'll look into lua testing now ) Thanks again for your help.

**Comment by dexterfhy** [comment]
Hi Justin, I added a test for jobstop to check if it returns 0 once data->is_rpc. 

I'm having difficulty figuring out how one would test if an error is actually printed (assuming it originally did, with `EMSG()`).. if one were to test that. 

I saw instances of `eq({'notification', 'stderr', {0, {'...'}}}, next_msg())` in other functions, but those check for stderr whereas `EMSG()` in `message.c` writes elsewhere specifically right?

**Comment by justinmk** [comment]
> having difficulty figuring out how one would test if an error is actually printed

In that case it would raise an error, then `nvim('command', ...` would fail. So it's already tested by your test.

Scheduling this for 0.5 since it's somewhat risky. Please ping next week if I forget about it.

**Comment by dexterfhy** [comment]
Got it, thanks for all your help!

File: test/functional/core/job_spec.lua [hunk 4]
```diff
@@ -875,6 +875,13 @@ describe('jobs', function()
     end)
   end)
 
+  it('jobstop fails silently and returns exit status', function()
+    nvim('command', "let j = jobstart(['cat'], {'rpc': v:true})")
+    neq(0, eval('j'))
+    nvim('command', "let h = jobstop(j)")
+    eq(0, eval('h'))
```
Comment (justinmk) [inline test/functional/core/job_spec.lua]: i assume you're still working on this? 0=success, but this should test a failed jobstop(). E.g. try stopping the same job twice.

File: src/nvim/eval.c [hunk 5]
```diff
@@ -12419,10 +12419,7 @@ static void f_jobstop(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     (void)channel_close(data->id, kChannelPartRpc, &error);
   }
   process_stop((Process *)&data->stream.proc);
-  rettv->vval.v_number = 1;
-  if (error) {
-    EMSG(error);
-  }
+  rettv->vval.v_number = error == NULL ? 0 : 1;
```
Comment (justinmk) [inline src/nvim/eval.c]: still needs to return the exit status? That would be in `data->stream.proc.status` or something like that.

**Comment by jfelchner** [comment]
@justinmk since I believe `termopen` uses the job infrastructure behind the scenes, would this PR fit into your recommended solution for this? https://github.com/neovim/neovim/issues/4713

**Comment by justinmk** [comment]
> @justinmk since I believe `termopen` uses the job infrastructure behind the scenes, would this PR fit into your recommended solution for this? #4713

@jfelchner good note, but I don't think so. This PR is only relevant when jobstart/termopen fails immediately.

File: src/nvim/eval.c [hunk 6]
```diff
@@ -12404,11 +12404,10 @@ static void f_jobstop(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 
   if (argvars[0].v_type != VAR_NUMBER) {
     // Only argument is the job id
-    EMSG(_(e_invarg));
     return;
```
Comment (justinmk) [inline src/nvim/eval.c]: why was this removed? If an invalid argument type is provided, it's not useful to just ignore it.

**Comment by justinmk** [comment]
Continuing at https://github.com/neovim/neovim/pull/11740
